File: bioreactor/src/main/java/com/cardioflex/bioreactor/MainActivity.java
package com.cardioflex.bioreactor;

import android.annotation.SuppressLint;
import android.os.Build;
import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import org.eclipse.milo.opcua.sdk.client.OpcUaClient;
import org.eclipse.milo.opcua.sdk.client.api.UaClient;
import org.eclipse.milo.opcua.stack.core.UaException;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.structured.ReferenceDescription;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import io.okandroid.OkAndroid;
import io.okandroid.opcua.OpcClient;

public class MainActivity extends AppCompatActivity {
    TextView textView;
    Button button;

    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        textView = findViewById(R.id.log_text);
        button = findViewById(R.id.test_btn);
        button.setOnClickListener(view -> {
            textView.setText("Done!  000000");
            try {
                OkAndroid.newThread().scheduleDirect(() -> {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        run();
                    }
                });
            } catch (Exception e) {
                e.printStackTrace();
                textView.setText(e.getLocalizedMessage());
            }
        });
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void run() {
        OkAndroid.mainThread().createWorker().schedule(() -> {
            textView.setText("Done!  0");
        });
        new OpcClient((client, future) -> {
            try {
                OkAndroid.mainThread().createWorker().schedule(() -> {
                    textView.setText("Done!  1");
                });
                System.out.println("Hello!");
                UaClient uaClient = client.connect().get();
                List<ReferenceDescription> descriptions = uaClient.getAddressSpace().browse(NodeId.parse(""));
                System.out.println(descriptions.size());
                future.complete(client);
                OkAndroid.mainThread().createWorker().schedule(() -> {
                    textView.setText("Done!  2");
                });
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }
}

File: motor/src/main/java/com/cardioflex/motor/MainActivity.kt
package com.cardioflex.motor

import android.os.Bundle
import android.widget.GridLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import io.okandroid.serial.SerialDevice

class MainActivity : AppCompatActivity() {
    lateinit var device: SerialDevice

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        title = "蠕动泵控制"
        init_modbus()
    }

    private fun init_modbus() {
        try {
            device = SerialDevice.newBuilder("/dev/ttyS8", 115200).build()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        val gridLayout: GridLayout = findViewById(R.id.control_container)
        for (index in IntArray(4) { it }) {
            val pane = ControlPane(this, device, index + 1)
            val params = GridLayout.LayoutParams()
            params.columnSpec = GridLayout.spec(index % 2)
            params.rowSpec = GridLayout.spec(index / 2)
//            pane.layoutParams = params
            gridLayout.addView(pane, params)
        }
//        gridLayout.addView(ControlPane(this, device, 0x01), 0)
//        gridLayout.addView(ControlPane(this, device, 0x02), 1)
//        gridLayout.addView(ControlPane(this, device, 0x03), 2)
//        gridLayout.addView(ControlPane(this, device, 0x04), 3)
    }
}